resources/tentative/todomvc/src/components/todo-item/todo-item.component.js

```diff
@@ -0,0 +1,194 @@
+        this.item = node.querySelector(".todo-item");
+        this.toggleLabel = node.querySelector(".toggle-todo-label");
+        this.toggleInput = node.querySelector(".toggle-todo-input");
+        this.todoText = node.querySelector(".todo-item-text");
+        this.todoButton = node.querySelector(".remove-todo-button");
+        this.editLabel = node.querySelector(".edit-todo-label");
+        this.editInput = node.querySelector(".edit-todo-input");
+        // create shadow dom
+        this.shadow = this.attachShadow({ mode: "open" });
+        // rtl support to target with styles
+        this.htmlDirection = document.querySelector("html").getAttribute("dir") || "ltr";
+        this.shadow.host.setAttribute("dir", this.htmlDirection);
+        // add shadow dom
+        this.shadow.append(node);
+        // listeners
+        this.keysListeners = [];
+        // bind event handlers
+        this.updateItem = this.updateItem.bind(this);
+        this.toggleItem = this.toggleItem.bind(this);
+        this.removeItem = this.removeItem.bind(this);
+        this.startEdit = this.startEdit.bind(this);
+        this.stopEdit = this.stopEdit.bind(this);
+        this.cancelEdit = this.cancelEdit.bind(this);
+    }
+
+    update(...args) {
+        [...args].forEach((argument) => {
+            switch (argument) {
+                case "id":
+                    if (this.id !== undefined) {
+                        this.item.id = `todo-item-${this.id}`;
+                        this.toggleLabel.setAttribute(
+                            "for",
+                            `toggle-todo-${this.id}`
+                        );
+                        this.toggleInput.id = `toggle-todo-${this.id}`;
+                        this.todoText.id = `update-todo-${this.id}`;
+                        this.todoButton.id = `remove-todo-${this.id}`;
+                        this.editLabel.setAttribute(
+                            "for",
+                            `edit-todo-${this.id}`
+                        );
+                        this.editInput.id = `edit-todo-${this.id}`;
+                    }
+                    break;
+                case "title":
+                    if (this.title !== undefined) {
+                        this.todoText.textContent = this.title;
+                        this.editInput.value = this.title;
+                    }
+                    break;
+                case "completed":
+                    this.toggleInput.checked = this.getAttribute("completed") === "true" ? true : false;
+                    break;
+            }
+        });
+    }
+
+    startEdit() {
+        this.item.classList.add("editing");
+        this.editInput.value = this.title;
+        this.editInput.focus();
+    }
+
+    stopEdit() {
+        this.item.classList.remove("editing");
+    }
+
+    cancelEdit() {
+        this.editInput.blur();
+    }
+
+    toggleItem() {
+        // update item first, before dispatch
```

Comment:
I see that from the code but why is that necessary?

Reply:
1.  The item updates its "completed" state   
2. The app listens for the "toggle-item" event and updates the attributes on topbar, list, bottombar.
3. The list checks the "completed" attribute to filter based on route (therefore I need the completed state to already be updated before the check)

Reply:
So I think it would be more useful for this comment to say 'The todo-list checks the "completed" attribute to filter based on route (therefore the completed state needs to already be updated before the check)'